### PR TITLE
Adding JFrog Distribution

### PR DESCRIPTION
**Pull request by jainishshah17**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: incubator/distribution/templates/distribution-deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        env:
+        - name: DEFAULT_JAVA_OPTS
+          value: '-Ddistribution.home={{ .Values.distribution.persistence.mountPath }} -Dcom.sun.management.jmxremote.authenticate=false
+                  -Dcom.sun.management.jmxremote.ssl=false -Duser.timezone=UTC
+                  -Xms{{ .Values.distribution.resources.requests.memory | trunc 1 }}g
+                  -Xmx{{ .Values.distribution.resources.limits.memory | trunc 1 }}g
+                  -Dspring.profiles.active=production'
+        - name: mongo_connectionString
+        {{- if .Values.mongodb.mongodbUsername }}
+          value: 'mongodb://{{ .Values.mongodb.db.distributionUser }}:{{ .Values.mongodb.db.distributionPassword }}@{{ .Release.Name }}-mongodb:27017/{{ .Values.mongodb.db.name }}'
```
Comment (rjkernick) [inline incubator/distribution/templates/distribution-deployment.yaml]: Best practice is usually to wrap up connection details in a secret.

File: incubator/distribution/templates/distributor-deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        imagePullPolicy: {{ .Values.distributor.image.imagePullPolicy }}
+        env:
+        - name: DEFAULT_JAVA_OPTS
+          value: '-Ddistribution.home={{ .Values.distributor.persistence.mountPath }} -Dfile.encoding=UTF8 -Dcom.sun.management.jmxremote.authenticate=false
+                  -Dcom.sun.management.jmxremote.ssl=false -Duser.timezone=UTC
+                  -Xms{{ .Values.distributor.resources.requests.memory | trunc 1 }}g
+                  -Xmx{{ .Values.distributor.resources.limits.memory | trunc 1 }}g
+                  -Dspring.profiles.active=production'
+        - name: redis_connectionString
+          value: 'redis://:{{ .Values.redis.redisPassword }}@{{ .Release.Name }}-redis:6379'
```
Comment (rjkernick) [inline incubator/distribution/templates/distributor-deployment.yaml]: Please also wrap up the connection details into a secret.

File: incubator/distribution/templates/hooks/create-user.yaml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+{{ toYaml .Values.nodeSelector | indent 8 }}
+      {{- end }}
+      restartPolicy: OnFailure
+      containers:
+      - name: post-install-job
+        image: "mvertes/alpine-mongo:{{ .Values.mongodb.image.tag }}-0"
+        command:
+        - 'sh'
+        - '-c'
+        - 'sh /scripts/setup.sh'
```
Comment (rjkernick) [inline incubator/distribution/templates/hooks/create-user.yaml]: Do you have issues with this running? Configmaps are mounted readonly.

**Review by rjkernick — CHANGES_REQUESTED** [review]

File: incubator/distribution/templates/hooks/create-user.yaml (hunk 3 above)
Comment (jainishshah17) [reply incubator/distribution/templates/hooks/create-user.yaml]: @rjkernick  It works for me on k8s 1.9.6.

File: incubator/distribution/templates/mongo-connection-secret.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  labels:
+    app: {{ template "distribution.name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+type: Opaque
+data:
+  {{- if .Values.mongodb.mongodbUsername }}
+  mongo_connectionString: {{ template "mongodb.url" . }}
+  audit_mongo_connectionString: {{ template "mongodb.audit.url" . }}
```
Comment (rjkernick) [inline incubator/distribution/templates/mongo-connection-secret.yaml]: Can probably remove the if/else here as they have the same body.

**Review by rjkernick — CHANGES_REQUESTED** [review]

File: incubator/distribution/templates/mongo-connection-secret.yaml (hunk 4 above)
Comment (jainishshah17) [reply incubator/distribution/templates/mongo-connection-secret.yaml]: @rjkernick done

File: incubator/distribution/templates/_helpers.tpl [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+*/}}
+{{- define "distributor.name" -}}
+{{- default .Chart.Name .Values.distributor.name .Values.nameOverride | trunc 63 | trimSuffix "-" -}}
+{{- end -}}
+
+{{/*
+Create a default fully qualified distribution name.
+We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
+*/}}
+{{- define "distribution.fullname" -}}
```
Comment (unguiculus) [inline incubator/distribution/templates/_helpers.tpl]: Why doesn't this allow a fullnameOverride?

File: incubator/distribution/templates/distribution-deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,111 @@
+apiVersion: extensions/v1beta1
```
Comment (unguiculus) [inline incubator/distribution/templates/distribution-deployment.yaml]: apps/v1beta2

File: incubator/distribution/templates/distribution-deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+kind: Deployment
+metadata:
+  name: {{ template "distribution.fullname" . }}
+  labels:
+    app: {{ template "distribution.name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version }}
```
Comment (unguiculus) [inline incubator/distribution/templates/distribution-deployment.yaml]: The latest `_helpers.tpl` you get with `helm create` has a `chart` partial. Please update accordingly.

File: incubator/distribution/templates/distribution-deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    component: {{ .Values.distribution.name }}
```
Comment (unguiculus) [inline incubator/distribution/templates/distribution-deployment.yaml]: Do component names have to be configurable?

File: incubator/distribution/templates/distributor-deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,80 @@
+apiVersion: extensions/v1beta1
```
Comment (unguiculus) [inline incubator/distribution/templates/distributor-deployment.yaml]: apps/v1beta2

File: incubator/distribution/values.yaml [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+      cpu: "250m"
+  ## Make sure the --wiredTigerCacheSizeGB is no more than half the memory limit!
+  ## This is critical to protect against OOMKill by Kubernetes!
+  mongodbExtraFlags:
+  - "--wiredTigerCacheSizeGB=1"
+  usePassword: false
+  db:
+    name: bintray
+    adminUser: admin
+    adminPassword: password
```
Comment (unguiculus) [inline incubator/distribution/values.yaml]: Passwords should be auto-generated by default.

File: incubator/distribution/values.yaml [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+##
+mongodb:
+  enabled: true
+  image:
+    tag: 3.6.3
+    pullPolicy: IfNotPresent
+  persistence:
+    enabled: true
+    size: 10Gi
+  resources:
```
Comment (unguiculus) [inline incubator/distribution/values.yaml]: Don't set defaults for resources. Comment them out.

File: incubator/distribution/values.yaml [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+    # storageClass: "-"
+
+distributor:
+  replicaCount: 1
+  name: distributor
+  image:
+    repository: "docker.bintray.io/jfrog/distribution-distributor"
+    version: 1.0.0
+    imagePullPolicy: IfNotPresent
+  token: "eyJ2ZXIiOiIyIiwidHlwIjoiSldUIiwiYWxnIjoiUlMyNTYiLCJraWQiOiJ6WWFWQjVFRlpkOXlpbWtCNkZTbDAtSWFhMHRPWldLX293c1BTVTNTZmdNIn0.eyJzdWIiOiJJTlRFUk5BTCIsInNjcCI6ImFwcGxpZWQtcGVybWlzc2lvbnNcL3N5c3RlbSIsImF1ZCI6ImpmZHNAMDFjN3pxd2FnOWNmNTAwd2FnOXZhcTF3cjkiLCJpc3MiOiJqZmRzQDAxYzd6cXdhZzljZjUwMHdhZzl2YXExd3I5IiwiZXhwIjoyNDY3NjI1MDczLCJpYXQiOjE1MjE1NDUwNzMsImp0aSI6IjE0ZjJjN2IxLTMzYzQtNDA0Yy04YTllLTJmZjU2OGFjODRlOSJ9.ACfT3pfu4wTTt1wOkyqwYCp9UUVWNwMLl3ZyuO-igDR9N_G-G3oMkuCFekAzgOrTcbGjrXTDMNwoRyYhHUs4wtYxM9yAA-u72EhVz10stL9B0d2JUOamc8xJaLvQDhL38TYiTjdgNVsy726qgxzX2xnx1M8VSp9Qnw7r2ZGBQ6LjrESIUZb8IXy-93oxYqDSkWvjOJZiNQXhJBgOzhzq3r-buaGgxOMoC7DduCctCT0FIxd1jrsZPUnpH3YGE374j7YGcGBdy7eQj-skM3YircJVgk6XFYJBKZBJ8pX3NLZznirG2m9QbhLCRhH1T6B4g-HZvy251wz3QBUdmzO6Cg"
```
Comment (unguiculus) [inline incubator/distribution/values.yaml]: You should not hard-code tokens. You may provide custom values for CI only if necessary: https://github.com/kubernetes/charts/blob/master/test/README.md#providing-custom-test-values

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/distribution/README.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+### Updating Distribution
+Once you have a new chart version, you can update your deployment with
+```
+helm upgrade distribution incubator/distribution
+```
+
+## Configuration
+
+The following table lists the configurable parameters of the distribution chart and their default values.
+
```
Comment (unguiculus) [inline incubator/distribution/README.md:42]: Could you format the table properly?

File: incubator/distribution/templates/distribution-deployment.yaml [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+          valueFrom:
+            secretKeyRef:
+              name: {{ template "distribution.fullname" . }}-redis-connection
+              key: redis_connectionString
+        - name: BT_ARTIFACTORY_URL
+          value: {{ .Values.distribution.env.artifactoryUrl | quote }}
+        - name: BT_SERVER_URL
+          value: {{ .Values.distribution.env.btServerUrl | quote }}
+        {{- if .Values.distribution.env.artifactoryEdge1Url }}
+        - name: artifactory_edge_1_url
```
Comment (unguiculus) [inline incubator/distribution/templates/distribution-deployment.yaml:76]: Just wondering. It's kind of uncommon to have env vars lowercase. If case doesn't matter, I'd suggest you make them all uppercase.

File: incubator/distribution/values.yaml [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+  #    memory: "2Gi"
+  #    cpu: "100m"
+  #  limits:
+  #    memory: "2Gi"
+  #    cpu: "250m"
+  ## Make sure the --wiredTigerCacheSizeGB is no more than half the memory limit!
+  ## This is critical to protect against OOMKill by Kubernetes!
+  mongodbExtraFlags:
+  - "--wiredTigerCacheSizeGB=1"
+  mongodbRootPassword: password
```
Comment (unguiculus) [inline incubator/distribution/values.yaml]: Don't specify default passwords. You could used the `required` function to display error messages and provide passwords for CI only.

See https://github.com/kubernetes/charts/blob/master/test/README.md#providing-custom-test-values

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/distribution/templates/distribution-deployment.yaml (hunk 14 above)
Comment (jainishshah17) [reply incubator/distribution/templates/distribution-deployment.yaml:76]: It's Case sensitive. Will get fixed with the future release of our product.

File: incubator/distribution/README.md (hunk 13 above)
Comment (jainishshah17) [reply incubator/distribution/README.md:42]: done

**Review by unguiculus — APPROVED** [review]

**Pull request merged**